### PR TITLE
Accept IApiConnection on AuthenticationApiClient & ManagementApiClient

### PR DESCRIPTION
**Pull request by damieng**
Provides a simple way to mock out the AuthenticationApiClient and ManagementApiClient underlying implementations without sprawling concerns or breaking changes when we add endpoints.

Related to #250

### CODE REVIEW

File: src/Auth0.AuthenticationApi/AuthenticationApiClient.cs [hunk 1]
```diff
@@ -406,11 +411,24 @@ public async Task UnlinkUserAsync(UnlinkUserRequest request)
         }
 
         /// <summary>
-        /// Disposes of any owned disposable resources such as the ApiConnection.
+        /// Dispose of any managed resources such as the <see cref="IApiConnection"/>.
+        /// </summary>
+        /// <param name="disposing">Whether to dispose of managed resources right now.</param>
+        protected virtual void Dispose(bool disposing)
```
Comment (joshcanhelp) [inline src/Auth0.AuthenticationApi/AuthenticationApiClient.cs]: I'm confused why this is necessary. Why would you call this with `disposing` set to false?

File: src/Auth0.AuthenticationApi/AuthenticationApiClient.cs [hunk 2]
```diff
@@ -17,10 +17,17 @@ namespace Auth0.AuthenticationApi
     public class AuthenticationApiClient : IAuthenticationApiClient, IDisposable
     {
         private readonly Uri _baseUri;
-        private readonly ApiConnection _apiConnection;
+        private IApiConnection _apiConnection;
```
Comment (joshcanhelp) [inline src/Auth0.AuthenticationApi/AuthenticationApiClient.cs]: Does `ApiConnection` extend `IApiConnection`? Or am I missing a different relationship here?

File: src/Auth0.AuthenticationApi/AuthenticationApiClient.cs (hunk 1 above)
Comment (damieng) [reply src/Auth0.AuthenticationApi/AuthenticationApiClient.cs]: It's the standard .NET pattern for disposing objects that can be inherited from.  VS suggests it strongly from the IDE.

File: src/Auth0.AuthenticationApi/AuthenticationApiClient.cs (hunk 2 above)
Comment (damieng) [reply src/Auth0.AuthenticationApi/AuthenticationApiClient.cs]: IApiConnection is the interface, ApiConnection is our concrete class.  By accepting an interface instead customers can create their own implementation that instead returns canned responses for their testing.

**Comment by damieng** [comment]
Going to do more work on this PR splitting `IApiConnection` to `IManagementApiConnection` and `IAuthenticationApiConnection` so that either side is easier  to mock and also combine the various overloads for different http methods into a single one with a parameter to further aid with mocking.

**Comment by damieng** [comment]
IApiConnection is a bad interface for AuthenticationClient, there should be two interfaces one for each. Rejecting in favor of an alternative (incoming).

**Pull request closed without merging**